### DISCUSSION TITLE
container-queries-plugin: Allow container and modifiers on same element

### DISCUSSION THREAD

**Discussion by Soviut** [tailwindlabs/tailwindcss#11713] [Ideas]
Currently, when using the `@tailwindcss/container-queries` plugin, a `@container` declaration must be on an element that wraps its "modifiers" (`@md`, `@xl`, etc.).

```html
<div class="@container">
  <ul class="grid gap-3 @md:grid-cols-2 @xl:grid-cols-3 @2xl:grid-cols-4">
```

However, this wrapping adds clutter to components.

It would be very convenient if that container declaration and modifiers could exist on the same element.

```html
<ul class="@container grid gap-3 @md:grid-cols-2 @xl:grid-cols-3 @2xl:grid-cols-4">
```

As far as I know, this shouldn't cause any issues.

**Comment by wongjn** [comment]
Container queries work by marking an element as the containment context that the container queries evaluate against. This is done by `container-type` and the `@container` class applies `container-type: inline-size`.

Container queries are evaluated against *parent* containment contexts and not themselves. So the container queries would not be evaluated for the same element as the containment context, only any *parent* containment contexts.

**Reply by Soviut** [reply]
I didn't realize `container-type` needed to be on a parent element. I supposed this helps to avoid infinite loops as the contents changes size.

It's unfortunate since changing the template grid columns, which affect that element's children, definitely feels like it should be possible this way.
